File: frontend/src/types/logistica.ts
import type { ID, ISODateString } from "./common";

export type SupplyCategory =
  | "water"
  | "food"
  | "medical"
  | "fuel"
  | "ppe"
  | "tools"
  | "other";

export const SUPPLY_LABEL: Record<SupplyCategory, string> = {
  water: "AGUA",
  food: "ALIMENTACIÓN",
  medical: "SANITARIO",
  fuel: "COMBUSTIBLE",
  ppe: "EPI",
  tools: "HERRAMIENTAS",
  other: "OTROS",
};

export type SupplyState = "ok" | "low" | "critical" | "out";

export interface SupplyItem {
  id: ID;
  category: SupplyCategory;
  name: string;
  unit: string;              // "L", "kg", "ud"
  stock: number;
  minStock: number;
  location: string;          // "Pabellón Robledo · Almacén 1"
  state: SupplyState;
  updatedAt: ISODateString;
}

export type VehicleKind =
  | "command-vehicle"
  | "transport-truck"
  | "supply-van"
  | "fuel-tank"
  | "off-road"
  | "bus";

export type VehicleState = "available" | "deployed" | "maintenance" | "out-of-service";

export interface Vehicle {
  id: ID;
  callSign: string;
  kind: VehicleKind;
  state: VehicleState;
  capacity: string;          // "8 plazas", "8000 L", "5 t"
  fuelPct: number;           // 0..100
  driver?: string;
  location: string;          // "Base Logística Robledo"
  updatedAt: ISODateString;
}

export type MachineryKind =
  | "generator"
  | "pump"
  | "lighting"
  | "compressor"
  | "crane"
  | "bulldozer";

export interface Machinery {
  id: ID;
  callSign: string;
  kind: MachineryKind;
  state: "operational" | "deployed" | "maintenance" | "down";
  location: string;
  notes?: string;
  updatedAt: ISODateString;
}

export type ServiceKind = "water" | "electricity" | "gas" | "telecom" | "sewer";

export const SERVICE_LABEL: Record<ServiceKind, string> = {
  water: "AGUA",
  electricity: "ELECTRICIDAD",
  gas: "GAS",
  telecom: "TELEFONÍA",
  sewer: "SANEAMIENTO",
};

export type ServiceStatus = "operational" | "degraded" | "outage" | "restoring";

export interface InfrastructureService {
  id: ID;
  kind: ServiceKind;
  area: string;              // zona afectada
  status: ServiceStatus;
  affectedPopulation: number;
  provider: string;
  notes?: string;
  estimatedRestore?: ISODateString;
  updatedAt: ISODateString;
}

export type LogisticsRequestStatus =
  | "pending"
  | "approved"
  | "in-transit"
  | "delivered"
  | "denied";

export type LogisticsRequestPriority = "critical" | "high" | "medium" | "low";

export interface LogisticsRequest {
  id: ID;
  emergencyId: ID;
  requestedBy: string;
  category: SupplyCategory | "vehicle" | "machinery" | "personnel";
  itemName: string;
  quantity: number;
  unit: string;
  priority: LogisticsRequestPriority;
  status: LogisticsRequestStatus;
  destination: string;
  requestedAt: ISODateString;
  decidedAt?: ISODateString;
  eta?: ISODateString;
  assignedVehicleId?: ID | null;
  notes?: string;
}

export type LogisticsActivityType =
  | "stock-updated"
  | "request-created"
  | "request-approved"
  | "request-denied"
  | "request-delivered"
  | "vehicle-assigned"
  | "service-status-changed"
  | "shelter-updated";

export interface LogisticsActivity {
  id: ID;
  type: LogisticsActivityType;
  message: string;
  performedBy: string;
  timestamp: ISODateString;
  refId?: ID;
}

File: backend/app/schemas/logistica.py
"""Schemas del área logística."""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class SuministroCreate(BaseModel):
    category: str = Field(min_length=1, max_length=16)
    name: str = Field(min_length=1, max_length=120)
    unit: str = Field(default="ud", max_length=20)
    stock: int = Field(default=0, ge=0)
    min_stock: int = Field(default=0, ge=0)
    location: str | None = Field(default=None, max_length=160)


class SuministroRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    emergencia_id: int
    category: str
    name: str
    unit: str
    stock: int
    min_stock: int
    location: str | None


class StockAjuste(BaseModel):
    delta: int


class SolicitudCreate(BaseModel):
    requested_by: str = Field(min_length=1, max_length=120)
    category: str = Field(min_length=1, max_length=20)
    item_name: str = Field(min_length=1, max_length=160)
    quantity: int = Field(default=1, ge=1)
    unit: str | None = Field(default=None, max_length=20)
    priority: str = Field(default="medium", pattern="^(critical|high|medium|low)$")
    destination: str | None = Field(default=None, max_length=160)


class SolicitudRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    emergencia_id: int
    requested_by: str
    category: str
    item_name: str
    quantity: int
    unit: str | None
    priority: str
    estado: str
    destination: str | None
    requested_at: datetime
    decided_at: datetime | None


class ServicioCreate(BaseModel):
    kind: str = Field(min_length=1, max_length=16)
    area: str = Field(min_length=1, max_length=160)
    affected_population: int = Field(default=0, ge=0)
    provider: str | None = Field(default=None, max_length=120)


class ServicioRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    emergencia_id: int
    kind: str
    area: str
    estado: str
    affected_population: int
    provider: str | None


class EstadoUpdate(BaseModel):
    estado: str = Field(min_length=1, max_length=12)
